packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,423 @@
+   */
+  const computeVstorageAbciRoute = (height, kind, path) =>
+    encodeURI(
+      `/abci_query?data=0x${encodeHex(
+        codecs[kind].request.toProto({ path }),
+      )}&height=${height}&path="${PATH_PREFIX}/${kind}"`,
+    );
+
+  /**
+   * @overload
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['CHILDREN'] }>} opts
+   * @returns {Promise<QueryChildrenResponse>}
+   *
+   * @overload
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['DATA'] }>} opts
+   * @returns {Promise<QueryDataResponse>}
+   */
+  /**
+   * @param {string} path
+   * @param {Partial<{ height: Height; kind: PATHS['CHILDREN'] | PATHS['DATA'] }>} [opts]
+   * @returns {Promise<QueryChildrenResponse | QueryDataResponse>}
+   */
+  const queryAbci = async (
+    path,
+    { height = 0n, kind = PATHS.CHILDREN } = {},
+  ) => {
+    await null;
+
+    const codec = codecs[kind];
+    /** @type {Response|undefined} */
+    let response;
+    const route = computeVstorageAbciRoute(height, kind, path);
+    // TODO: implement some sort of load distribution
+    // mechanism to smartly choose an rpc address
+    const rpcAddress = config.rpcAddrs.find(Boolean);
+
+    const url = rpcAddress + route;
+
+    try {
+      response = await fetch(url, { keepalive: true });
+    } catch (err) {
+      console.error(err);
+      throw makeVstorageError({
+        errorMessage: err.message,
+        height,
+        kind,
+        path,
+        rpcAddress,
+      });
+    }
+
+    if (!response.ok)
+      throw makeVstorageError({
+        errorMessage: await response.text(),
+        height,
+        kind,
+        path,
+        rpcAddress,
+      });
+
+    const { result } = /** @type {AbciResponse} */ (await response.json());
+
+    if (result.response?.code !== 0)
+      throw makeVstorageError({
+        errorMessage: `Error code '${result.response.code}', Error message: '${result.response.log}'`,
+        height,
+        kind,
+        path,
+        rpcAddress,
+      });
+
+    return codec.response.decode(decodeBase64(result.response.value));
+  };
+
+  return { queryAbci };
+};
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @returns {StreamTopic<T>}
+ */
+const makeStreamTopic = ({ queryClient }, path) => {
+  /**
+   * @param {Height} blockHeight
+   * @param {StreamCell['values']} values
+   */
+  function* allUpdatesFromCell(blockHeight, values) {
+    for (const value of values)
+      yield /** @type {Update<T>} */ ({
+        blockHeight,
+        value,
+      });
+  }
+
+  /**
+   * @param {Height} [minimum]
+   * @param {Height} [maximum]
+   */
+  async function* iterate(minimum, maximum) {
+    /** @type {Array<Update<T>>} */
+    let values = [];
+
+    for await (const value of reverseIterate(maximum, minimum))
+      values = [value, ...values];
+
+    for (const value of values) yield value;
+  }
```

Comment:
It's simpler to collect and then reverse, and probably also more efficient:
```suggestion
  async function* iterate(minimum, maximum) {
    /** @type {Array<Update<T>>} */
    const updates = [];

    for await (const update of reverseIterate(maximum, minimum)) {
      updates.push(update);
    }

    yield* updates.reverse();
  }
```

Reply:
Also, note that agoric-sdk convention is to define `reverseIterate` before this use of it.